miio/gateway.py

```diff
@@ -126,58 +117,48 @@ def light(self) -> "GatewayLight":
         """Return light control interface."""
         return self._light
 
-    @command()
+    @property
     def devices(self):
-        """Return list of devices."""
-        # from https://github.com/aholstenson/miio/issues/26
-        devices_raw = self.send("get_device_prop", ["lumi.0", "device_list"])
-        devices = [
-            SubDevice(self, *devices_raw[x : x + 5])  # noqa: E203
-            for x in range(0, len(devices_raw), 5)
-        ]
-
-        return devices
+        """Return a list of the already discovered devices."""
+        return self._devices
 
-    @command(click.argument("sid"), click.argument("property"))
-    def get_device_prop(self, sid, property):
+    @command()
+    def discover_devices(self):
+        """Discovers SubDevices and returns a list of the discovered devices."""
+        # from https://github.com/aholstenson/miio/issues/26
+        devices_raw = self.get_gateway_prop("device_list")
+        self._devices = []
+        thismodule = sys.modules[__name__]
+        
+        for x in range(0, len(devices_raw), 5):
```

Comment:
This is very convoluted way compared to the previous version...

The better way would be to create a mapping between subdevice types and their implementation classes. Accessing `sys.modules` and `getattr` are usually code smells better to be avoided wherever possible.

Reply:
I will improve on this but will take a few days since I don't have much time now

Reply:
I now use a dictionary lookup instead, unfortunately I cannot define the dictionary at the top of the file since the class defenitions need to be executed first before this dict can be defined.

Reply:
Looks much better! Hmm, maybe it will make sense to refactor the gateway to be contained in its own directory, and defining the separate devices in separate files:
```
miio/gateway/gateway.py
miio/gateway/aqaramagnet.py
miio/gateway/light.py
miio/gateway/zigbee.py
```
etc. Or at least have the subdevice classes in their own file?